core/src/display_object/movie_clip.rs

```diff
@@ -2839,6 +2832,141 @@ impl<'gc> TInteractiveObject<'gc> for MovieClip<'gc> {
         None
     }
 
+    fn mouse_pick_avm2(
+        &self,
+        context: &mut UpdateContext<'_, 'gc>,
+        point: (Twips, Twips),
+        require_button_mode: bool,
+    ) -> Avm2MousePick<'gc> {
+        if self.visible() {
+            let this: InteractiveObject<'gc> = (*self).into();
+
+            if let Some(masker) = self.masker() {
+                if !masker.hit_test_shape(context, point, HitTestOptions::SKIP_INVISIBLE) {
+                    return Avm2MousePick::Miss;
+                }
+            }
+
+            // In AVM2, mouse_enabled should only impact the ability to select the current clip
+            // but it should still be possible to select any children where child.mouse_enabled() is
+            // true.
+            // InteractiveObject.mouseEnabled:
+            // "Any children of this instance on the display list are not affected."
+            if self.mouse_enabled() && self.world_bounds().contains(point) {
+                // This MovieClip operates in "button mode" if it has a mouse handler,
+                // either via on(..) or via property mc.onRelease, etc.
+                let is_button_mode = self.is_button_mode(context);
+
+                if is_button_mode {
+                    let mut options = HitTestOptions::SKIP_INVISIBLE;
+                    options.set(HitTestOptions::SKIP_MASK, self.maskee().is_none());
+                    if self.hit_test_shape(context, point, options) {
+                        return Avm2MousePick::Hit(this);
+                    }
+                }
+            }
+
+            // Maybe we could skip recursing down at all if !world_bounds.contains(point),
+            // but a child button can have an invisible hit area outside the parent's bounds.
+            let mut options = HitTestOptions::SKIP_INVISIBLE;
+            options.set(HitTestOptions::SKIP_MASK, self.maskee().is_none());
+
+            let mut found_propagate = None;
+
+            let mut clip_layers = self
+                .iter_render_list()
+                .flat_map(|child| {
+                    if child.clip_depth() > 0 {
+                        // Note - we intentionally use 'child.depth()' here insteado
+                        // of the position in the render list - this matches Flash's
+                        // behavior. The 'depth' value comes from the PlaceObject tag
+                        Some((child, (child.depth() + 1)..=(child.clip_depth())))
+                    } else {
+                        None
+                    }
+                })
+                .rev()
+                .peekable();
+
+            // Interactive children run first, followed by non-interactive children.
+            // Depth is considered within each group.
+            let (interactive, non_interactive): (Vec<_>, Vec<_>) = self
```

Comment:
This can be done without collecting Vecs (two iterators with `filter()`), right? Or are there lifetime issues?